airbyte-commons/src/main/java/io/airbyte/commons/util/MoreIterators.java

```diff
@@ -0,0 +1,236 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.commons.util;
+
+import com.google.common.collect.AbstractIterator;
+import io.airbyte.commons.concurrency.SafeVoidCallable;
+import io.airbyte.commons.concurrency.VoidCallable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public class MoreIterators {
+
+  /**
+   * Create an iterator from elements
+   *
+   * @param elements element to put in iterator
+   * @param <T> type
+   * @return iterator with all elements
+   */
+  public static <T> Iterator<T> of(T... elements) {
+    return Arrays.asList(elements).iterator();
+  }
+
+  /**
+   * Create a list from an iterator
+   *
+   * @param iterator iterator to convert
+   * @param <T> type
+   * @return list
+   */
+  public static <T> List<T> toList(Iterator<T> iterator) {
+    final List<T> list = new ArrayList<>();
+    while (iterator.hasNext()) {
+      list.add(iterator.next());
+    }
+    return list;
+  }
+
+  /**
+   * Create a set from an iterator
+   *
+   * @param iterator iterator to convert
+   * @param <T> type
+   * @return set
+   */
+  public static <T> Set<T> toSet(Iterator<T> iterator) {
+    final Set<T> set = new HashSet<>();
+    while (iterator.hasNext()) {
+      set.add(iterator.next());
+    }
+    return set;
+  }
+
+  /**
+   * Create an iterator that only executes the given supplier after the first invocation of hasNext.
+   * When fully consumed it will close the input stream.
+   *
+   * @param streamSupplier stream to be supplied lazily
+   * @param <T> type
+   * @return lazy, auto closing iterator
+   */
+  public static <T> Iterator<T> streamSupplierToLazyIterator(Supplier<Stream<T>> streamSupplier) {
+    return iteratorSupplierToLazyIterator(() -> autoCloseIterator(streamSupplier.get()));
+  }
+
+  /**
+   * Create an iterator that only executes the given supplier after the first invocation of hasNext.
+   *
+   * @param iteratorSupplier iterator to be supplied lazily
+   * @param <T> type
+   * @return lazy iterator
+   */
+  public static <T> Iterator<T> iteratorSupplierToLazyIterator(Supplier<Iterator<T>> iteratorSupplier) {
+    return new LazyIterator<>(iteratorSupplier);
+  }
+
+  /**
+   * Creates an iterator from an iterator. The new iterator when fully consumed will execute the
+   * onClose function.
+   *
+   * @param iterator input iterator
+   * @param onClose on close function to be executed when iterator is completely consumed
+   * @param <T> type
+   * @return auto closing iterator
+   */
+  public static <T> Iterator<T> autoCloseIterator(Iterator<T> iterator, SafeVoidCallable onClose) {
```

Comment:
This is a VERY dangerous function. It should not return an `Iterator`, it should return an AutoCloseIterator. Otherwise there is absolutely no way to release resources but to consume the iterator to the end.

Reply:
removed.